### DISCUSSION TITLE
How to convert a radical SMILES with explicit Hydrogen to XYZ?

### DISCUSSION THREAD

**Discussion by Quappas** [rdkit/rdkit#4703] [Q&A] [answered]
After digging through the rdkit documentation, I wrote this piece of code:

```
from rdkit import Chem
from rdkit.Chem import AllChem

mol = Chem.MolFromSmiles("[H]C([H])([H])C")
print("Num: " + mol.GetNumAtoms())
AllChem.EmbedMolecule(mol)
print(Chem.MolToXYZBlock(mol))
```

The output is:

```
Num: 2
2

C      0.752009    0.000000    0.000000
C     -0.752009    0.000000    0.000000
```

I would like to get the Hydrogen atoms that are explicitly included in the SMILES string to be included in the XYZ output. `Chem.AddHs()` does not help.

**Comment by dehaenw** [comment]
Greetings @Quappas . Perhaps the explicitOnly parameter of addHs could work for you case?
```python
from rdkit import Chem
from rdkit.Chem import AllChem

mol = Chem.MolFromSmiles("[H]C([H])([H])C")
print("Num: ",mol.GetNumAtoms())
mol = Chem.AddHs(mol,explicitOnly=True)
AllChem.EmbedMolecule(mol)
print(Chem.MolToXYZBlock(mol))
```
gives
```
Num:  2
5

C     -0.069306    0.010527   -0.014392
C      1.432647    0.044259    0.076850
H     -0.361726   -0.804537   -0.710062
H     -0.473480    0.995891   -0.318981
H     -0.528135   -0.246141    0.966585
```

**Reply by Quappas** [reply]
Thank you, that works, but after replacing the final `C` with the proper `[C]`, `AddHs()` without parameters seems to work as well. I assume `AddHs()` gets confused when there is both explicit and accidentally implicit Hydrogen.

**Comment by greglandrum** [comment, marked as answer]
Hi @Quappas,
The way you tell the RDKit to preserve the Hs you are providing in a SMILES is with the `removeHs` field of a `SmilesParserParams` object.

Here's a short example:
```
In [2]: ps = Chem.SmilesParserParams()

In [3]: ps.removeHs = False

In [4]: m = Chem.MolFromSmiles('[H]C([H])([H])C')

In [5]: m.GetNumAtoms()
Out[5]: 2

In [6]: m = Chem.MolFromSmiles('[H]C([H])([H])C',ps)

In [7]: m.GetNumAtoms()
Out[7]: 5
```

**Reply by greglandrum** [reply]
By the way, the correct way to do a radical SMILES for the molecule you are working with is: `C[C]` or `[H]C([H])([H])[C]`. The square brackets around the second C atom indicates that it doesn't have any implicit Hs, which means it must be a radical.

**Reply by Quappas** [reply]
Thank you for the heads-up! I actually used a combination of `RWMol.AddAtom()`, `RWMol().AddBond()` and `MolToSmiles()` to construct incorrect SMILES like `[H]C([H])([H])C`, by explicitly adding all atoms to an empty molecule. After setting `allHsExplicit=True` in `MolToSmiles()`, I got the nicer `[C]C`. For the correct radical SMILES, `AddHs()` without parameters works just as well as for non-radicals.
